### DISCUSSION TITLE
[sqlparser] How to create a DEFAULT constraint programatically

### DISCUSSION THREAD

**Discussion by cachapa** [simolus3/drift#2243] [Q&A] [answered]
My goal is to intercept `CREATE TABLE` statements and inject columns transparently in the SQL statement.
One of the columns needs a `DEFAULT 0` constraint, but I can't seem to create one due to `NumericToken` requiring a `FileSpan` instance which is abstract and it seems its implementation is private.

Here's my code:

```dart
  Future<void> execute(String sql, [List<Object?>? args]) {
    final result = SqlEngine().parse(sql);

    if (result.rootNode is CreateTableStatement) {
      final statement = result.rootNode as CreateTableStatement;
      final createStatement = CreateTableStatement(
        tableName: statement.tableName,
        columns: [
          ...statement.columns,
          ColumnDefinition(
            columnName: 'is_deleted',
            typeName: 'INTEGER',
            constraints: [Default(null, NumericLiteral(0, NumericToken(--???--)))],
          ),
        ],
        tableConstraints: statement.tableConstraints,
        ifNotExists: statement.ifNotExists,
        isStrict: statement.isStrict,
        withoutRowId: statement.withoutRowId,
      );

      sql = createStatement.toSql();
    }

    return _db.execute(sql, args);
  }
```

Is there any way to add the constraint that's not running String.replace through the resulting sql statement?

**Comment by simolus3** [comment, marked as answer]
Interesting approach! In tests for `sqlparser`, I'm doing this:

https://github.com/simolus3/drift/blob/a98890b1c96d2f58840d58dfd02df76d54082f7b/sqlparser/test/parser/utils.dart#L63-L65
https://github.com/simolus3/drift/blob/a98890b1c96d2f58840d58dfd02df76d54082f7b/sqlparser/test/parser/utils.dart#L9-L13

Then I can use `NumericLiteral(0, token(TokenType.numberLiteral))`. Requiring a token reference is not great though, I'll just change that.

**Reply by simolus3** [reply]
I've dropped the token as a parameter in 6e173e8c083665c56e472eff6657d16d1056120e

**Comment by cachapa** [comment]
Cool, thanks for the quick turnaround!

I've tested it locally and it works now. I had the same issue in other places (intercepting inserts and updates) and it's working there as well.
